File: go/compatibility.go
package workhorse

import (
	"context"
	"errors"
	"fmt"
	"sync"
)

// CompatibilityCode identifies why a client must refuse a mutation.
type CompatibilityCode string

const (
	SchemaNotInstalled   CompatibilityCode = "schema-not-installed"
	SchemaTooOld         CompatibilityCode = "schema-too-old"
	SchemaTooNew         CompatibilityCode = "schema-too-new"
	ClientProtocolTooOld CompatibilityCode = "client-protocol-too-old"
	ClientProtocolTooNew CompatibilityCode = "client-protocol-too-new"
)

// CompatibilityError refuses a mutation against an incompatible SQL protocol.
type CompatibilityError struct {
	Code CompatibilityCode
}

func (err *CompatibilityError) Error() string {
	return fmt.Sprintf("SQL protocol compatibility check refused mutation: %s", err.Code)
}

// Is matches compatibility errors by refusal code.
func (err *CompatibilityError) Is(target error) bool {
	other, ok := target.(*CompatibilityError)
	return ok && other != nil && err.Code == other.Code
}

// CheckCompatibility compares an installed schema and client protocol version.
//
// The client declares a floor and no ceiling. Inside a major line a migration only adds, so a
// schema newer than the one this build was compiled against still carries every function it calls.
// The ceiling comes from servedProtocolVersions, which the installed schema declares in
// workhorse.protocol_version: a major release drops the client protocols it stops serving.
func CheckCompatibility(installedSchemaVersion *int, clientProtocolVersion int, servedProtocolVersions []int) error {
	var code CompatibilityCode
	switch {
	case installedSchemaVersion == nil:
		code = SchemaNotInstalled
	case *installedSchemaVersion < minimumSchemaVersion:
		code = SchemaTooOld
	case clientProtocolVersion < minimumProtocolVersion:
		code = ClientProtocolTooOld
	case clientProtocolVersion > maximumProtocolVersion:
		code = ClientProtocolTooNew
	default:
		served, oldest := servedProtocol(servedProtocolVersions, clientProtocolVersion)
		switch {
		case served:
			return nil
		case clientProtocolVersion < oldest:
			// The database crossed a major boundary and stopped answering this client.
			code = SchemaTooNew
		default:
			// The database has not been migrated far enough to answer this client yet.
			code = SchemaTooOld
		}
	}
	return &CompatibilityError{Code: code}
}

// servedProtocol reports whether the installed schema answers this client, and the oldest protocol
// it still answers. An empty declaration is treated as serving every client, because a schema that
// records nothing has made no statement to enforce.
func servedProtocol(servedProtocolVersions []int, clientProtocolVersion int) (bool, int) {
	if len(servedProtocolVersions) == 0 {
		return true, clientProtocolVersion
	}
	oldest := servedProtocolVersions[0]
	for _, version := range servedProtocolVersions {
		if version < oldest {
			oldest = version
		}
		if version == clientProtocolVersion {
			return true, oldest
		}
	}
	return false, oldest
}

// AssertSchemaCompatible reads the installed schema on every call. One statement returns both the
// schema version and the client protocols the schema declares it serves, so the check stays one
// round trip.
func AssertSchemaCompatible(ctx context.Context, executor Executor) error {
	rows, err := executor.Query(ctx, internalStatementRegistry[compatibilityStateStatement])
	if err != nil {
		if hasSQLState(err, "42P01", "3F000") {
			return &CompatibilityError{Code: SchemaNotInstalled}
		}
		return err
	}

	var schemaVersions []int
	var served []int
	for _, row := range rows {
		version, ok := integer(row["version"])
		if !ok {
			continue
		}
		switch row[compatibilityKindColumn] {
		case compatibilityKindSchema:
			schemaVersions = append(schemaVersions, version)
		case compatibilityKindProtocol:
			served = append(served, version)
		}
	}

	var installed *int
	if len(schemaVersions) == 1 {
		installed = &schemaVersions[0]
	}
	return CheckCompatibility(installed, ProtocolVersion, served)
}

// CachedCompatibilityCheck runs one compatibility query and reuses its result.
type CachedCompatibilityCheck struct {
	executor Executor
	once     sync.Once
	err      error
}

// NewCachedCompatibilityCheck builds the opt-in one-shot gate for hot worker loops.
func NewCachedCompatibilityCheck(executor Executor) *CachedCompatibilityCheck {
	return &CachedCompatibilityCheck{executor: executor}
}

// Assert returns the result of the first compatibility query for every call.
func (check *CachedCompatibilityCheck) Assert(ctx context.Context) error {
	check.once.Do(func() {
		check.err = AssertSchemaCompatible(ctx, check.executor)
	})
	return check.err
}

type sqlStateCarrier interface {
	SQLState() string
}

func hasSQLState(err error, states ...string) bool {
	var carrier sqlStateCarrier
	if !errors.As(err, &carrier) {
		return false
	}
	for _, state := range states {
		if carrier.SQLState() == state {
			return true
		}
	}
	return false
}

func integer(value any) (int, bool) {
	switch value := value.(type) {
	case int:
		return value, true
	case int16:
		return int(value), true
	case int32:
		return int(value), true
	case int64:
		return int(value), true
	default:
		return 0, false
	}
}

File: rust/src/compatibility.rs
//! The startup schema compatibility check every SDK runs before its first mutation.
use std::fmt;

use crate::queue::Executor;
use crate::sql_catalogue_generated as sql;
use crate::Error;

/// Why PostgreSQL and this client cannot work together.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum CompatibilityCode {
    SchemaNotInstalled,
    SchemaTooOld,
    SchemaTooNew,
    ClientProtocolTooOld,
    ClientProtocolTooNew,
}

impl CompatibilityCode {
    /// The protocol's stable refusal code, for example `schema-too-old`.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::SchemaNotInstalled => "schema-not-installed",
            Self::SchemaTooOld => "schema-too-old",
            Self::SchemaTooNew => "schema-too-new",
            Self::ClientProtocolTooOld => "client-protocol-too-old",
            Self::ClientProtocolTooNew => "client-protocol-too-new",
        }
    }
}

impl fmt::Display for CompatibilityCode {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.as_str())
    }
}

/// The installed schema version and the protocol versions PostgreSQL serves.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct CompatibilityState {
    /// `None` when the schema is absent or its version table is ambiguous.
    pub installed_schema_version: Option<i32>,
    pub served_protocol_versions: Vec<i32>,
}

/// Decides compatibility from observed state without touching PostgreSQL.
pub fn check_compatibility(
    installed_schema_version: Option<i32>,
    client_protocol_version: i32,
    served_protocol_versions: &[i32],
) -> Result<(), CompatibilityCode> {
    let Some(installed) = installed_schema_version else {
        return Err(CompatibilityCode::SchemaNotInstalled);
    };
    if installed < sql::MINIMUM_SCHEMA_VERSION {
        return Err(CompatibilityCode::SchemaTooOld);
    }
    if client_protocol_version < sql::MINIMUM_PROTOCOL_VERSION {
        return Err(CompatibilityCode::ClientProtocolTooOld);
    }
    if client_protocol_version > sql::MAXIMUM_PROTOCOL_VERSION {
        return Err(CompatibilityCode::ClientProtocolTooNew);
    }
    if served_protocol_versions.is_empty()
        || served_protocol_versions.contains(&client_protocol_version)
    {
        return Ok(());
    }
    let oldest = served_protocol_versions.iter().copied().min().unwrap_or(0);
    Err(if client_protocol_version < oldest {
        CompatibilityCode::SchemaTooNew
    } else {
        CompatibilityCode::SchemaTooOld
    })
}

/// Reads the compatibility state; a missing schema reads as not installed.
pub async fn read_compatibility_state<E: Executor>(
    executor: &E,
) -> Result<CompatibilityState, Error> {
    let rows = match executor.rows(sql::COMPATIBILITY_STATE, &[]).await {
        Ok(rows) => rows,
        Err(error) if matches!(error.sqlstate(), Some("42P01" | "3F000")) => {
            return Ok(CompatibilityState::default());
        }
        Err(error) => return Err(error),
    };
    let mut schema = Vec::new();
    let mut state = CompatibilityState::default();
    for row in rows {
        let version: i32 = row.try_get("version")?;
        match row.try_get::<_, &str>("kind")? {
            "schema" => schema.push(version),
            "protocol" => state.served_protocol_versions.push(version),
            _ => {}
        }
    }
    if let [version] = schema[..] {
        state.installed_schema_version = Some(version);
    }
    Ok(state)
}

/// Refuses with [`Error::Compatibility`] unless this client can use the installed schema.
pub async fn assert_schema_compatible<E: Executor>(executor: &E) -> Result<(), Error> {
    let state = read_compatibility_state(executor).await?;
    check_compatibility(
        state.installed_schema_version,
        sql::CLIENT_PROTOCOL_VERSION,
        &state.served_protocol_versions,
    )
    .map_err(|code| Error::Compatibility { code })
}
